working/augmentation-libraries/feature-specification.md

```diff
@@ -321,82 +322,150 @@ doesn't, it must repeat the type from the augmented definition__.
+    occur only as the function of a function invocation, which means
+    it is immediately followed by one, or no, type argument list
+    and then an argument list.
+    _Tear offs are not possible, neither are member accesses like
+    `augmented.call()`._
+    The expression `augmented` must not occur in an optional parameter
+    default value expressions _It is a reserved word in such expressions,
+    but it currently cannot be used._
+    The term <code>'augmented' \<argumentPart\></code> words parses
+    as a `<primary>` expression.
+    Evaluating the function invocation, after having evaluated the
+    argument list, invokes the augmented function with the provided
+    type arguments and evaluated argument list, and evaluates to the
+    returned value.
+    Type inference of <Code>augmented\<_typeArgs_\>(_args_)<code>
+    is performed as a function invocation with the function signature
+    of the augmented declaration.
+
+*   **Augmenting operators**: Operators are methods with special
+    invocation syntax. When augmenting an operator declaration, `augmented`
+    refers to the augmented operator *method*, and it must occur only
+    as the function of a function invocation. For example, when augmenting
     `operator +` you could use `augmented(1)` to call the augmented
-    operator, and when augmenting `operator []=` you would use the
-    `augmented(key, value)` syntax.
-
-    *   Note that `augmented` in such an augmenting operator method body is
-        not an expression by itself, and cannot be used to tear off the
-        augmented operator method. Similar to `super`, it is a syntactic
-        form which can only be used in limited ways.
+    operator, when augmenting `operator []=` you would use the
+    `augmented(key, value)` syntax, and when augmenting `operator -()`,
+    you would use `augmented()`.
+    Parsing, evaluation and type inference is handled like a
+    function augmentation.
 
 *   **Augmenting enum values**: When augmenting an enum value, `augmented`
-    has no meaning and is not allowed.
+    has no meaning and is not allowed to occur in constructor argument
+    expressions.
+
+*   **Augmenting constructors**:
+    *   **Non-redirecting generative constructor**:
+        The term `augmented` has no meaning, and must not occur in
+        optional parameter default value expressions, initializer list
+        entry expressions (whether initializing or asserts),
+        super-constructor invocation argument expressions, or in the
+        body. _Non-redirecting generative constructors automatically
+        invoke their augmented constructor after executing their
+        initializer list, and will execute their body after that of
+        their augmented constructor._
+    *   **Non-redirecting Factory constructor**: An `augmented` in the
+        constructor body refers to the augmented non-redirecting factory
+        constructor. It must occur only as the function of a (non-generic)
```

Comment:
Interesting restriction regarding it being a non-generic invocation. That might be a good limitation, but I will note that this program works today.

```dart
void main() {
  print(A<num>.foo());
}

class A<T> {
  A();

  factory A.foo() {
    return A<int>() as A<T>;
  }
}
```

We might want to consider also blocking this for regular function augmentations if we are going to block it for factory constructors?

Reply:
I really hope we get generic constructors at some point, so I want to reserve `augmented<int>()` to mean that, and `<int>augmented()` doesn't look great. Or parse at all.

Reply:
SGTM